### PR TITLE
Add AntiAffinity to Loki and Memcached

### PR DESCRIPTION
**Pull request by aminesnow**
This PR introduces a flag to add `AntiAffinity` to Loki component manifests and Loki caches' manifests.

### CODE REVIEW

**Comment by aminesnow** [comment]
CC @periklis

File: configuration/components/loki.libsonnet [hunk 1]
```diff
@@ -48,6 +48,7 @@ local defaults = {
     compactor: {
       withLivenessProbe: true,
       withReadinessProbe: true,
+      withAffinity: true,
```
Comment (douglascamata) [inline configuration/components/loki.libsonnet:51]: Sounds like this could be called `withPodAntiAffinity`, as that's effectively what's added?

**Review by douglascamata — APPROVED** [review]
LGTM. @periklis might also want to have a look at this.

File: configuration/components/loki.libsonnet [hunk 2]
```diff
@@ -652,6 +662,29 @@ function(params) {
       if std.length(config.resources) > 0
     },
 
+  local affinity(component, namespace) = {
+    podAntiAffinity: {
+      preferredDuringSchedulingIgnoredDuringExecution: [
+        {
+          weight: 100,
```
Comment (periklis) [inline configuration/components/loki.libsonnet:669]: Why is the weight here `100`. Does this have a meaning?

File: configuration/components/loki.libsonnet [hunk 3]
```diff
@@ -652,6 +662,29 @@ function(params) {
...
+          podAffinityTerm: {
+            topologyKey: 'kubernetes.io/hostname',
+            namespaces: [namespace],
```
Comment (periklis) [inline configuration/components/loki.libsonnet:672]: Can you elaborate a bit in the description on what happens if the target cluster will not have enough distinct nodes, e.g. for 3 ingesters schedule each on a different node?

**Review by periklis — CHANGES_REQUESTED** [review]
LGTM, however I would rather put `withPodAntiAffinity: false` in the default configuration because this node placement directive requires checking the node capacity upfront. Thus the user should opt-in by making a explicit choice.

File: configuration/components/loki.libsonnet (hunk 3 above)
Comment (aminesnow) [reply configuration/components/loki.libsonnet:672]: Here we use `preferredDuringSchedulingIgnoredDuringExecution`, which is a soft matcher and is ignored when unable to be fulfilled.

**Review by periklis — COMMENTED** [review]
What is the reason to prefer `preferredDuringSchedulingIgnoredDuringExecution` over `requiredDuringSchedulingIgnoredDuringExecution`? Why can't the user make a choice?

FYI, rendering the upstream [loki mixing](https://github.com/grafana/loki/tree/44f1d8d7f6b1fb9b7fac32e913bd6517d15bad3f/production/ksonnet/loki) tells me that they use `requiredDuringSchedulingIgnoredDuringExecution` for ingesters, rulers, query-frontend and query-scheduler pods. This means they apply strict scheduling for at least the components which:
- Either have impact/ensure replication (i.e. ingester)
- Or spread the load on executing queries (query-frontend, query-scheduler)
- Or spread the load on executing rules

WDYT to get in touch with the Loki team and ask for advice?

**Comment by aminesnow** [comment]
> What is the reason to prefer `preferredDuringSchedulingIgnoredDuringExecution` over `requiredDuringSchedulingIgnoredDuringExecution`? Why can't the user make a choice?
> 
> FYI, rendering the upstream [loki mixing](https://github.com/grafana/loki/tree/44f1d8d7f6b1fb9b7fac32e913bd6517d15bad3f/production/ksonnet/loki) tells me that they use `requiredDuringSchedulingIgnoredDuringExecution` for ingesters, rulers, query-frontend and query-scheduler pods. This means they apply strict scheduling for at least the components which:
> 
>     * Either have impact/ensure replication (i.e. ingester)
> 
>     * Or spread the load on executing queries (query-frontend, query-scheduler)
> 
>     * Or spread the load on executing rules
> 
> 
> WDYT to get in touch with the Loki team and ask for advice?

The reason is because I don't want this constraint to be hard, ie, prevent the deployment when the condition is not fulfilled. 
I don't see what advice could the upstream give me, since we don't run nearly anywhere near the same conditions. We should focus on what we want to achieve for our use case. 
For me, this is about making one step towards spreading the pods around the nodes without making it impossible to deploy them. If this is not the goal here then I should close this PR and we should spike this.

**Review by periklis — APPROVED** [review]

**Pull request merged**